rfcs/001-deterministic-states-and-caching.md

```diff
@@ -0,0 +1,716 @@
+
+```tsx
+const [{ data, error, loading }, getBalance] = useBalance({
+  addressOrName: name || address
+  formatUnits: 'gwei',
+  skip: !name && !address
+})
+```
+
+I feel like if we flip the logic, and introduce an `enabled` argument, this would make DX a bit easier to follow & reason about:
+
+```tsx
+const [{ data, error, loading }, getBalance] = useBalance({
+  addressOrName: name || address
+  formatUnits: 'gwei',
+  enabled: Boolean(name || address)
+})
+```
+
+By default, if `enabled` is not provided, then it is truthy & it still fetches on mount.
+
+## React Query & `Provider`
+
+This implementation will use [React Query](https://react-query.tanstack.com/), which means we will need to provide a query client (via react-query's `QueryClientProvider`) within wagmi's `Provider`.
+
+## Storage persistence
+
+I think maybe we should have storage persistence an opt-in feature, that people can toggle on from the `Provider`. Storage persistence is available via [react-query's `persistQueryClient`](https://react-query.tanstack.com/plugins/persistQueryClient)
+
+### Usage
+
+```tsx
+import { Provider, createPersistor } from 'wagmi';
+
+const localStoragePersistor = createPersistor({ storage: window.localStorage })
+
+const App = () => (
+  <Provider persistor={localStoragePersistor}>
+    <YourRoutes />
+  </Provider>
+)
+```
+
+## Read hooks
+
+The following hooks will see a common set of changes to their APIs:
+
+### Arguments
+
+The following set of **optional** arguments will be added to the read hooks:
+
+- `cacheTime: number`: The time (in ms) that the data remains in the cache, cached data is displayed initially to the user while a background fetch is being performed. Defaults to `5 * 60 * 1000` (5 minutes).
+- `staleTime: number`: The time (in ms) after which the data is considered stale, and a background refetch is needed. Defaults to `0`.
+- `onSuccess: (data: TData) => void`: A callback function to be invoked when the data is successfully fetched.
+- `onError: (error: TError) => void`: A callback function to be invoked when an error has been thrown.
+- `onSettled: (data?: TData, error?: TError) => void`: A callback function to be invoked when the data has either been fetched successfully or not (kinda like `finally`).
+- `suspense: boolean`: Enable the hook to use suspense
+
+### Return values
+
+The `loading` flag will be removed in favour of the following set of deterministic states:
+
+- `isIdle`: A flag to indicate that the promise has not yet been invoked, and data has not been loaded yet.
+- `isLoading`: A flag to indicate that the data is loading for the first time
+- `isReloading`: A flag to indicate that the data is loading in the background, and showing cached success/error data in the meantime (meaining this flag can be truthy when isSuccess/isError is truthy).
+- `isSuccess`: A flag to indicate the data has been fetched successfully.
+- `isError`: A flag to indicate that an error was thrown upon fetching the account
+
```

Comment:
Can we also expose `status` too? Think it's a useful prop for folks that don't like booleans.

Reply:
Yes, great idea.